File: src/main/java/org/sylfra/idea/plugins/remotesynchronizer/ui/config/ConfigPanel.java
package org.sylfra.idea.plugins.remotesynchronizer.ui.config;

import com.intellij.ide.plugins.PluginManagerCore;
import com.intellij.openapi.extensions.PluginId;
import com.intellij.openapi.project.Project;
import com.intellij.ui.components.JBLabel;
import com.intellij.ui.components.JBPanel;
import com.intellij.ui.components.JBTabbedPane;
import org.imyuyu.idea.plugins.filesync.FileSyncPlugin;
import org.sylfra.idea.plugins.remotesynchronizer.model.Config;
import org.sylfra.idea.plugins.remotesynchronizer.ui.config.panes.GeneralPane;
import org.sylfra.idea.plugins.remotesynchronizer.ui.config.panes.IConfigPane;
import org.sylfra.idea.plugins.remotesynchronizer.ui.config.panes.LogPane;
import org.sylfra.idea.plugins.remotesynchronizer.ui.config.panes.TargetsPane;
import org.sylfra.idea.plugins.remotesynchronizer.utils.ConfigPathsManager;

import javax.swing.*;
import java.awt.*;
import java.util.Objects;

/**
 * Configuration panel
 */
public class ConfigPanel extends JPanel
{
  private Project currentProject;
  private IConfigPane[] configPanes;

  public ConfigPanel(FileSyncPlugin plugin)
  {
    configPanes = new IConfigPane[]{
      new TargetsPane(plugin.pathManager),
      new GeneralPane(plugin.project, plugin.pathManager),
      new LogPane()
    };

    buildUI(plugin.pathManager);
    currentProject = plugin.project;
  }

  public Project getCurrentProject()
  {
    return currentProject;
  }

  public boolean isModified(Config config)
  {
    for (IConfigPane configPane : configPanes)
    {
      if (configPane.isModified(config))
      {
        return true;
      }
    }

    return false;
  }

  public void reset(Config config)
  {
    for (IConfigPane configPane : configPanes)
    {
      configPane.reset(config);
    }
  }

  public void apply(Config config)
  {
    for (IConfigPane configPane : configPanes)
    {
      configPane.apply(config);
    }
  }

  private void buildUI(ConfigPathsManager pathsManager)
  {
    JPanel pnHeader = createHeaderPanel();

    JBTabbedPane mainPane = new JBTabbedPane();

    for (IConfigPane configPane : configPanes)
    {
      configPane.buildUI(pathsManager);
      mainPane.addTab(configPane.getTitle(), (Component) configPane);
    }

    setLayout(new BorderLayout());
    add(pnHeader, BorderLayout.NORTH);
    add(mainPane, BorderLayout.CENTER);
  }

  private JPanel createHeaderPanel()
  {
    JPanel pnImportExport = new JBPanel<>();

    pnImportExport.add(new JButton(new ActionsHolder.ImportAction(this)));
    pnImportExport.add(new JButton(new ActionsHolder.ExportAction(this)));

    JPanel panel = new JPanel(new BorderLayout());
    panel.add(pnImportExport, BorderLayout.CENTER);
    panel.add(new JBLabel("version " +
      Objects.requireNonNull(PluginManagerCore.getPlugin(PluginId.getId(FileSyncPlugin.PLUGIN_NAME))).getVersion()),
      BorderLayout.EAST);
    return panel;
  }
}

File: src/main/kotlin/org/imyuyu/idea/plugins/filesync/ui/config/ConfigPanel.kt
package org.imyuyu.idea.plugins.filesync.ui.config

import com.intellij.ide.plugins.PluginManagerCore
import com.intellij.openapi.extensions.PluginId
import com.intellij.openapi.project.Project
import com.intellij.ui.components.JBLabel
import com.intellij.ui.components.JBPanel
import com.intellij.ui.components.JBTabbedPane
import org.imyuyu.idea.plugins.filesync.FileSyncPlugin
import org.imyuyu.idea.plugins.filesync.model.Config
import org.imyuyu.idea.plugins.filesync.utils.ConfigPathsManager
import org.imyuyu.idea.plugins.filesync.ui.config.panes.GeneralPane
import org.imyuyu.idea.plugins.filesync.ui.config.panes.IConfigPane
import org.imyuyu.idea.plugins.filesync.ui.config.panes.LogPane
import org.imyuyu.idea.plugins.filesync.ui.config.panes.TargetsPane
import java.awt.BorderLayout
import java.awt.Component
import java.util.*
import javax.swing.JButton
import javax.swing.JPanel

/**
 * Configuration panel
 */
class ConfigPanel(plugin: FileSyncPlugin) : JPanel() {
    val currentProject: Project
    private val configPanes: Array<IConfigPane>

    init {
        configPanes = arrayOf(
            TargetsPane(plugin.pathManager),
            GeneralPane(plugin.project, plugin.pathManager),
            LogPane()
        )
        buildUI(plugin.pathManager)
        currentProject = plugin.project
    }

    fun isModified(config: Config): Boolean {
        for (configPane in configPanes) {
            if (configPane.isModified(config)) {
                return true
            }
        }
        return false
    }

    fun reset(config: Config) {
        for (configPane in configPanes) {
            configPane.reset(config)
        }
    }

    fun apply(config: Config) {
        for (configPane in configPanes) {
            configPane.apply(config)
        }
    }

    private fun buildUI(pathsManager: ConfigPathsManager) {
        val pnHeader = createHeaderPanel()
        val mainPane = JBTabbedPane()
        for (configPane in configPanes) {
            configPane.buildUI(pathsManager)
            mainPane.addTab(configPane.title, configPane as Component)
        }
        layout = BorderLayout()
        add(pnHeader, BorderLayout.NORTH)
        add(mainPane, BorderLayout.CENTER)
    }

    private fun createHeaderPanel(): JPanel {
        val pnImportExport: JPanel = JBPanel<JBPanel<*>>()
        pnImportExport.add(JButton(ActionsHolder.ImportAction(this)))
        pnImportExport.add(JButton(ActionsHolder.ExportAction(this)))
        val panel = JPanel(BorderLayout())
        panel.add(pnImportExport, BorderLayout.CENTER)
        panel.add(
            JBLabel(
                "version " +
                        Objects.requireNonNull(PluginManagerCore.getPlugin(PluginId.getId(FileSyncPlugin.PLUGIN_NAME)))?.version
            ),
            BorderLayout.EAST
        )
        return panel
    }
}
